### PR TITLE
Support deleting in batch requests to avoid AWS API request limits

### LINKED ISSUE #21: RequestLimitExceeded error when resources are a lot

### PR DESCRIPTION
**Pull request by tonerdo**
Fixes #21

### CODE REVIEW

File: aws/aws.go [hunk 1]
```diff
@@ -177,8 +191,21 @@ func NukeAllResources(account *AwsAccountResources, regions []string) error {
 
 		resourcesInRegion := account.Resources[region]
 		for _, resources := range resourcesInRegion.Resources {
-			if err := resources.Nuke(session); err != nil {
-				return errors.WithStackTrace(err)
+			if len(resources.ResourceIdentifiers()) >= 500 {
+				logging.Logger.Infoln("Terminating Resources in batches")
+				batches := split(resources.ResourceIdentifiers(), 200)
```
Comment (brikis98) [inline aws/aws.go]: Where does `500` and `200` come from? These are "magic" numbers and should be extracted into constants with some explanation above them (e.g., AWS throttles after XXX API calls...).

File: aws/aws.go [hunk 2]
```diff
@@ -177,8 +191,21 @@ func NukeAllResources(account *AwsAccountResources, regions []string) error {
 
 		resourcesInRegion := account.Resources[region]
 		for _, resources := range resourcesInRegion.Resources {
-			if err := resources.Nuke(session); err != nil {
-				return errors.WithStackTrace(err)
+			if len(resources.ResourceIdentifiers()) >= 500 {
+				logging.Logger.Infoln("Terminating Resources in batches")
```
Comment (brikis98) [inline aws/aws.go]: "There are more than XXX resources to terminate (YYY total), so terminating resources in batches"

File: aws/aws.go (hunk 1 above)
Comment (brikis98) [reply aws/aws.go]: In fact, each resource type should probably have a `MaxBatchSize()` function.

File: aws/aws.go [hunk 3]
```diff
@@ -177,8 +191,21 @@ func NukeAllResources(account *AwsAccountResources, regions []string) error {
...
+				batches := split(resources.ResourceIdentifiers(), 200)
+
+				for _, batch := range batches {
+					if err := resources.NukeBatch(session, batch); err != nil {
```
Comment (brikis98) [inline aws/aws.go]: It feels a bit weird to have a nearly identical `NukeBatch` and `Nuke` method in each resource type. Perhaps there should solely be a `Nuke` method that takes a list of identifiers as a parameter?

File: aws/aws.go (hunk 1 above)
Comment (tonerdo) [reply aws/aws.go]: Sorry about the "magic" numbers. Was trying stuff out and forgot to abstract it.

Comment (brikis98) [reply aws/aws.go]: AWS is vague on what the limits are: https://docs.aws.amazon.com/AWSEC2/latest/APIReference/query-api-troubleshooting.html. However, they vary per service and even per region. See also https://stackoverflow.com/questions/23728591/ec2-api-throttling-limits. 

I suspect we will need the flexibility to use different limits with different services. Also, it may turn out simple batching doesn't work and we need an [exponential back off](https://docs.aws.amazon.com/general/latest/gr/api-retries.html), but let's try the simple version first.

Comment (tonerdo) [reply aws/aws.go]: Ok, Gotcha. Thanks for the links

File: aws/ami_types.go [hunk 4]
```diff
@@ -21,9 +21,13 @@ func (image AMIs) ResourceIdentifiers() []string {
 	return image.ImageIds
 }
 
+func (image AMIs) MaxBatchSize() int {
+	return 200
```
Comment (brikis98) [inline aws/ami_types.go]: Please put a comment that this is a best guess!

Comment (brikis98) [reply aws/ami_types.go]: Same for all the other impls of this method

File: aws/aws.go [hunk 5]
```diff
@@ -41,6 +41,20 @@ func getRandomRegion() string {
 	return allRegions[randIndex]
 }
 
+func split(identifiers []string, limit int) [][]string {
```
Comment (brikis98) [inline aws/aws.go]: Please add a bunch of unit tests for this method to check boundary conditions: empty list, 1 item list, odd number of items, even number of items, small limit, large limit, etc.

File: aws/aws.go [hunk 6]
```diff
@@ -166,6 +180,8 @@ func GetAllResources(regions []string, excludedRegions []string, excludeAfter ti
 
 // NukeAllResources - Nukes all aws resources
 func NukeAllResources(account *AwsAccountResources, regions []string) error {
+	// Resources more than 500 in number could cause AWS to throttle api calls
```
Comment (brikis98) [inline aws/aws.go]: Again, please mention that 500 is a rule of thumb based on experience, but the actual number will vary based on resource type, region, etc.

File: aws/aws.go [hunk 7]
```diff
@@ -177,8 +193,23 @@ func NukeAllResources(account *AwsAccountResources, regions []string) error {
 
 		resourcesInRegion := account.Resources[region]
 		for _, resources := range resourcesInRegion.Resources {
-			if err := resources.Nuke(session); err != nil {
-				return errors.WithStackTrace(err)
+			length := len(resources.ResourceIdentifiers())
+			if length >= MaxRequestSize {
```
Comment (brikis98) [inline aws/aws.go]: Actually, it doesn't seem like you should even do this check. Instead, always batch things. If there are fewer items than the `MaxBatchSize`, it will all end up in one batch anyway.

Comment (tonerdo) [reply aws/aws.go]: I'll remove this and up the batch size of all resources to 500
